From 2be7933539f053e224691b3a29f1cdeb8a7d6541 Mon Sep 17 00:00:00 2001
From: Jianwei Fan <jianwei.fan@rock-chips.com>
Date: Mon, 20 May 2024 13:57:30 +0800
Subject: [PATCH 294/305] media: i2c: rk628: only disable i2s data out only
 when disabling i2s out

Type: Fix
Redmine ID: N/A
Associated modifications: N/A
Test: N/A

Change-Id: I96ae219e1e0c1338120547a43726e759884f0a7c
Signed-off-by: Shunhua Lan <lsh@rock-chips.com>
Signed-off-by: Guochun Huang <hero.huang@rock-chips.com>
---
 drivers/media/i2c/rk628/rk628_hdmirx.c | 35 +++++++++++++++-----------
 drivers/media/i2c/rk628/rk628_hdmirx.h |  2 ++
 2 files changed, 22 insertions(+), 15 deletions(-)

diff --git a/drivers/media/i2c/rk628/rk628_hdmirx.c b/drivers/media/i2c/rk628/rk628_hdmirx.c
index 5ba290a11da0e7..e78b06fedf3a98 100644
--- a/drivers/media/i2c/rk628/rk628_hdmirx.c
+++ b/drivers/media/i2c/rk628/rk628_hdmirx.c
@@ -47,6 +47,7 @@ struct rk628_audioinfo {
 	bool i2s_enabled_default;
 	bool i2s_enabled;
 	int debug;
+	bool sample_flat;
 	bool fifo_ints_en;
 	bool ctsn_ints_en;
 	bool audio_present;
@@ -440,12 +441,16 @@ static void rk628_csi_delayed_work_audio_v2(struct work_struct *work)
 	}
 	audio_state->pre_state = fifo_status;
 
-	rk628_i2c_read(rk628, HDMI_RX_AUD_SPARE, &sample_flat);
-	sample_flat = sample_flat & AUDS_MAS_SAMPLE_FLAT;
-	if (!sample_flat)
-		rk628_i2c_update_bits(rk628, GRF_SYSTEM_CON0, SW_I2S_DATA_OEN_MASK, SW_I2S_DATA_OEN(0));
-	else
-		rk628_i2c_update_bits(rk628, GRF_SYSTEM_CON0, SW_I2S_DATA_OEN_MASK, SW_I2S_DATA_OEN(1));
+	if (aif->i2s_enabled) {
+		rk628_i2c_read(rk628, HDMI_RX_AUD_SPARE, &sample_flat);
+		sample_flat = !!(sample_flat & AUDS_MAS_SAMPLE_FLAT);
+		if (sample_flat != aif->sample_flat) {
+			dev_info(rk628->dev, "audio sample flat change to %d\n", sample_flat);
+			rk628_i2c_write(aif->rk628, HDMI_RX_AUD_SAO_CTRL, I2S_LPCM_BPCUV(0) | I2S_32_16(1) |
+					(sample_flat ? I2S_DATA_ENABLE_BITS(0xf) : I2S_DATA_ENABLE_BITS(0)));
+			aif->sample_flat = sample_flat;
+		}
+	}
 
 exit:
 	schedule_delayed_work(&aif->delayed_work_audio, msecs_to_jiffies(delay));
@@ -611,17 +616,16 @@ void rk628_hdmirx_audio_i2s_ctrl(HAUDINFO info, bool enable)
 {
 	struct rk628_audioinfo *aif = (struct rk628_audioinfo *)info;
 
-	if (enable == aif->i2s_enabled)
+	if (enable == aif->i2s_enabled || aif->i2s_enabled_default)
 		return;
-	if (enable) {
+	if (enable && !aif->sample_flat) {
 		rk628_i2c_write(aif->rk628, HDMI_RX_AUD_SAO_CTRL,
-				I2S_LPCM_BPCUV(0) |
-				I2S_32_16(1));
+				I2S_LPCM_BPCUV(0) | I2S_32_16(1) |
+				I2S_DATA_ENABLE_BITS(0));
 	} else {
 		rk628_i2c_write(aif->rk628, HDMI_RX_AUD_SAO_CTRL,
-				I2S_LPCM_BPCUV(0) |
-				I2S_32_16(1) |
-				I2S_ENABLE_BITS(0x3f));
+				I2S_LPCM_BPCUV(0) | I2S_32_16(1) |
+				I2S_DATA_ENABLE_BITS(0xf));
 	}
 	aif->i2s_enabled = enable;
 }
@@ -651,6 +655,7 @@ void rk628_hdmirx_audio_setup(HAUDINFO info)
 	aif->audio_state.init_state = INIT_FIFO_STATE*4;
 	aif->audio_state.fifo_int = false;
 	aif->audio_state.audio_enable = false;
+	aif->sample_flat = false;
 	aif->fifo_ints_en = false;
 	aif->ctsn_ints_en = false;
 	aif->i2s_enabled = false;
@@ -686,8 +691,8 @@ void rk628_hdmirx_audio_setup(HAUDINFO info)
 			AFIF_SUBPACKETS(1));
 	rk628_i2c_write(aif->rk628, HDMI_RX_AUD_SAO_CTRL,
 			I2S_LPCM_BPCUV(0) |
-			I2S_32_16(1)|
-			(aif->i2s_enabled_default ? 0 : I2S_ENABLE_BITS(0x3f)));
+			I2S_32_16(1) |
+			(aif->i2s_enabled_default ? 0 : I2S_DATA_ENABLE_BITS(0xf)));
 	aif->i2s_enabled = aif->i2s_enabled_default;
 	rk628_i2c_write(aif->rk628, HDMI_RX_AUD_MUTE_CTRL,
 			APPLY_INT_MUTE(0)	|
diff --git a/drivers/media/i2c/rk628/rk628_hdmirx.h b/drivers/media/i2c/rk628/rk628_hdmirx.h
index de294d3287c363..c1f307078dae7a 100644
--- a/drivers/media/i2c/rk628/rk628_hdmirx.h
+++ b/drivers/media/i2c/rk628/rk628_hdmirx.h
@@ -189,6 +189,8 @@
 #define HDMI_RX_AUD_SAO_CTRL		(HDMI_RX_BASE + 0x0260)
 #define I2S_ENABLE_BITS_MASK		GENMASK(10, 5)
 #define I2S_ENABLE_BITS(x)		UPDATE(x, 10, 5)
+#define I2S_CLK_ENABLE_BITS(x)		UPDATE(x, 10, 9)
+#define I2S_DATA_ENABLE_BITS(x)		UPDATE(x, 8, 5)
 #define I2S_LPCM_BPCUV_MASK		BIT(11)
 #define I2S_LPCM_BPCUV(x)		UPDATE(x, 11, 11)
 #define I2S_32_16_MASK			BIT(0)
-- 
2.17.1

